v2/datastream-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/DataStreamToSpannerShardedMigrationIT.java

```diff
@@ -0,0 +1,400 @@
+      instance.tearDownBase();
+    }
+    ResourceManagerUtils.cleanResources(spannerResourceManager, pubsubResourceManager);
+  }
+
+  @Test
+  public void multiShardMigration() {
+    // Construct a ChainedConditionCheck with 4 stages.
+    // 1. Send initial wave of events
+    // 2. Wait on Spanner to have events
+    ChainedConditionCheck conditionCheck =
+        ChainedConditionCheck.builder(
+                List.of(
+                    uploadDataStreamFile(
+                        jobInfo1,
+                        TABLE,
+                        "Users-backfill-logical-shard1.jsonl",
+                        "DataStreamToSpannerShardedMigrationIT/Users-backfill-logical-shard1.jsonl"),
+                    uploadDataStreamFile(
+                        jobInfo1,
+                        TABLE,
+                        "Users-backfill-logical-shard2.jsonl",
+                        "DataStreamToSpannerShardedMigrationIT/Users-backfill-logical-shard2.jsonl"),
+                    uploadDataStreamFile(
+                        jobInfo1,
+                        TABLE,
+                        "Users-cdc-shard1.jsonl",
+                        "DataStreamToSpannerShardedMigrationIT/Users-cdc-shard1.jsonl")))
+            .build();
+
+    // Wait for conditions
+    PipelineOperator.Result result =
+        pipelineOperator()
+            .waitForCondition(createConfig(jobInfo1, Duration.ofMinutes(8)), conditionCheck);
+
+    // Assert Conditions
+    assertThatResult(result).meetsConditions();
+
+    conditionCheck =
+        ChainedConditionCheck.builder(
+                List.of(
+                    uploadDataStreamFile(
+                        jobInfo2,
+                        TABLE,
+                        "Users-backfill-logical-shard3.jsonl",
+                        "DataStreamToSpannerShardedMigrationIT/Users-backfill-logical-shard3.jsonl"),
+                    uploadDataStreamFile(
+                        jobInfo2,
+                        TABLE,
+                        "Users-backfill-logical-shard4.jsonl",
+                        "DataStreamToSpannerShardedMigrationIT/Users-backfill-logical-shard4.jsonl"),
+                    uploadDataStreamFile(
+                        jobInfo2,
+                        TABLE,
+                        "Users-cdc-shard2.jsonl",
+                        "DataStreamToSpannerShardedMigrationIT/Users-cdc-shard2.jsonl")))
+            .build();
+
+    result =
+        pipelineOperator()
+            .waitForCondition(createConfig(jobInfo2, Duration.ofMinutes(8)), conditionCheck);
+    assertThatResult(result).meetsConditions();
+
+    ConditionCheck rowsConditionCheck =
+        SpannerRowsCheck.builder(spannerResourceManager, TABLE)
+            .setMinRows(12)
+            .setMaxRows(12)
+            .build();
+    result =
+        pipelineOperator()
+            .waitForCondition(createConfig(jobInfo1, Duration.ofMinutes(10)), rowsConditionCheck);
```

Comment:
Does this mean it waits for 10 minutes for the check, and both the Dataflow pipelines would be running in parallel during this time?

Reply:
Yes.